### PR TITLE
#141 - New scenarios for large_transaction_test (& w/ GC)

### PR DESCRIPTION
**Pull request by amlatyrngom**
I noticed a failure while running the tests. Here is the corresponding log:

> Running main() from gtest_main.cc
[==========] Running 5 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 5 tests from LargeGCTests
[ RUN      ] LargeGCTests.MixedReadWriteWithGC
[       OK ] LargeGCTests.MixedReadWriteWithGC (12383 ms)
[ RUN      ] LargeGCTests.HighAbortRateWithGC
/home/amlatyr/Code/terrier/test/util/transaction_test_util.cpp:246: Failure
Value of: StorageTestUtil::ProjectionListEqual(layout_, entry.second, it->second)
  Actual: false
Expected: true
[  FAILED  ] LargeGCTests.HighAbortRateWithGC (7657 ms)
[ RUN      ] LargeGCTests.HighAbortRateHighThreadWithGC
[       OK ] LargeGCTests.HighAbortRateHighThreadWithGC (7041 ms)
[ RUN      ] LargeGCTests.TPCCWithGC
[       OK ] LargeGCTests.TPCCWithGC (16982 ms)
[ RUN      ] LargeGCTests.TPCCHighThreadWithGC
[       OK ] LargeGCTests.TPCCHighThreadWithGC (18133 ms)
[----------] 5 tests from LargeGCTests (62196 ms total)
[----------] Global test environment tear-down
[==========] 5 tests from 1 test case ran. (62196 ms total)
[  PASSED  ] 4 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] LargeGCTests.HighAbortRateWithGC
 1 FAILED TEST

I am unable to reproduce it.

### CODE REVIEW

File: test/storage/large_garbage_collector_test.cpp [hunk 1]
```diff
@@ -67,4 +67,115 @@ TEST_F(LargeGCTests, MixedReadWriteWithGC) {
...
+  const uint32_t num_concurrent_txns = TestThreadPool::HardwareConcurrency();
+  for (uint32_t iteration = 0; iteration < num_iterations; iteration++) {
+    LargeTransactionTestObject tested(max_columns, initial_table_size, txn_length, update_select_ratio, &block_store_,
+                                      &buffer_pool_, &generator_, true, true);
+    StartGC(tested.GetTxnManager(), 10);
+    for (uint32_t batch = 0; batch * batch_size < num_txns; batch++) {
+      auto result = tested.SimulateOltp(batch_size, num_concurrent_txns);
+      paused_ = true;
+      tested.CheckReadsCorrect(&result.first);
+      using namespace std;
```
Comment (mbutrovich) [inline test/storage/large_garbage_collector_test.cpp]: Linter doesn't like this. Not sure it'll change anything to remove this anyway.

File: test/storage/large_garbage_collector_test.cpp [hunk 2]
```diff
@@ -67,4 +67,115 @@ TEST_F(LargeGCTests, MixedReadWriteWithGC) {
...
+  const uint32_t num_concurrent_txns = 2*TestThreadPool::HardwareConcurrency();
+  for (uint32_t iteration = 0; iteration < num_iterations; iteration++) {
+    LargeTransactionTestObject tested(max_columns, initial_table_size, txn_length, update_select_ratio, &block_store_,
+                                      &buffer_pool_, &generator_, true, true);
+    StartGC(tested.GetTxnManager(), 10);
+    for (uint32_t batch = 0; batch * batch_size < num_txns; batch++) {
+      auto result = tested.SimulateOltp(batch_size, num_concurrent_txns);
+      paused_ = true;
+      tested.CheckReadsCorrect(&result.first);
+      using namespace std;
```
Comment (mbutrovich) [inline test/storage/large_garbage_collector_test.cpp]: See previous comment.

File: test/storage/large_garbage_collector_test.cpp [hunk 3]
```diff
@@ -67,4 +67,115 @@ TEST_F(LargeGCTests, MixedReadWriteWithGC) {
...
+// This test is similar to the previous one, but with a higher ration of updates
+// and longer transactions leading to more aborts.
+// NOLINTNEXTLINE
+TEST_F(LargeGCTests, HighAbortRateWithGC) {
+  const uint32_t num_iterations = 10;
+  const uint16_t max_columns = 2;
+  const uint32_t initial_table_size = 1000;
+  const uint32_t txn_length = 40;
+  const uint32_t num_txns = 1000;
+  const uint32_t batch_size = 200;
```
Comment (mbutrovich) [inline test/storage/large_garbage_collector_test.cpp]: Let's try to consolidate all of the shared test parameters in the test harness itself.

File: test/storage/large_garbage_collector_test.cpp [hunk 4]
```diff
@@ -67,4 +67,93 @@ TEST_F(LargeGCTests, MixedReadWriteWithGC) {
     EndGC();
   }
 }
+
+// This test is similar to the previous one, but with a higher ration of updates
```
Comment (tli2) [inline test/storage/large_garbage_collector_test.cpp]: ratio?

File: test/storage/large_garbage_collector_test.cpp [hunk 5]
```diff
@@ -67,4 +67,93 @@ TEST_F(LargeGCTests, MixedReadWriteWithGC) {
...
+      tested.CheckReadsCorrect(&result.first);
+      for (auto w : result.first) delete w;
+      for (auto w : result.second) delete w;
+      paused_ = false;
+    }
+    EndGC();
+  }
+}
+
+// This test attempts simulate a TPC-C-like scenario.
```
Comment (tli2) [inline test/storage/large_garbage_collector_test.cpp]: attempts to

**Comment by tli2** [comment]
@amlatyrngom Can you fix the merge conflict real quick?

**Comment by mbutrovich** [comment]
I'll clean up the merge conflict and make one more pass on this right now.

**Comment by mbutrovich** [comment]
I can't get it the segfault to reproduce for LargeTransactionTests.HighAbortRateHighThread scenario (same test but without GC) but I did just get...
```
/Users/matt/git/terrier/test/util/transaction_test_util.cpp:246: FailureValue of: StorageTestUtil::ProjectionListEqual(layout_, entry.second, it->second)
  Actual: false
Expected: true
```
... after running that test enough instances. So I'm not convinced it's just a GC issue. It might be a race in the abort logic somewhere that manifests faster with GC on.

**Comment by mbutrovich** [comment]
This looks good but is currently blocked by 2 issues:
1) llvm@6 home-brew target is busted so we can't get Mac coverage on CI.
2) there's a bug in the GC (that these new test scenarios caught, nice!) when operating on the same version chain as an Abort. We'll fix that first and then merge this in.

**Pull request merged**